README.md
# lokka-transport-http

Isomorphic HTTP Transport Layer for [Lokka](https://github.com/kadirahq/lokka)

---

This is a [graphql-express](https://github.com/graphql/express-graphql) compatible transport layer for [Lokka](https://github.com/kadirahq/lokka).

## Basic Usage

Install the package:

```
npm i --save lokka-transport-http
npm i --save lokka
```

This is how to send request to Facebook's [SWAPI GraphQL Demo](http://graphql-swapi.parseapp.com/).

```js
import HttpTransport from 'lokka-transport-http';
const transport = new HttpTransport('http://graphql-swapi.parseapp.com/');
transport.send(`
    {
      allFilms {
        films {
          title
        }
      }
    }
`).then(response => {
    console.log(JSON.stringify(response, null, 2));
});
```

## Send Custom Headers

It's possible to send custom headers like this:

```js
const headers = {
    'my-headers': 'some-value'
};
const transport = new HttpTransport('/graphql', {headers});
```

## Authentication

This package does not handle authentication information for you. But it'll let you interact with your app's existing authentication mechanism.

* If you already have an authorized cookie, it'll be sent with the HTTP request. (supports CORS)
* You can also set a custom `Authorization` [header]((https://www.npmjs.com/package/basic-auth-header) to implement [basic-auth](https://www.npmjs.com/package/basic-auth) support.

## Error Handling

By default it will create and throw a new `Error` object using the first GraphQL error. Error handling can be customized with the `handleErrors` option. Check the deafult error handler in `lib/index.js` for an example.


CHANGELOG.md
# Change Log

## v1.6.1
31 October 2017

Fixes [#18](https://github.com/kadirahq/lokka-transport-http/issues/18).
That causes due to a typo came with `v1.6.0`.

## v1.6.0
27 October 2017

* Support custom error handlers [PR17](https://github.com/kadirahq/lokka-transport-http/pull/17)

## v1.5.0

* Add the data object received to the Error object [PR9](https://github.com/kadirahq/lokka-transport-http/pull/9)
* optional CORS credentials support. [PR11](https://github.com/kadirahq/lokka-transport-http/pull/11)

## v1.4.1

Fix the way how we load fetch. Now it's work properly on safari as well. See [PR16](https://github.com/kadirahq/lokka-transport-http/pull/16)

## v1.4.0

* Remove isomorphic-fetch and handle isomorphic stuff overself. This is because isomorphic-fetch doesn't work properly in node 0.10.x

## v1.3.2

* Some code cleanup and potentially fix: [#3](https://github.com/kadirahq/lokka-transport-http/issues/3)

## v1.3.0
* Add react native support. See: [1598111](https://github.com/kadirahq/lokka-transport-http/tree/15981118287d7f72ba019937f3c0fbd0af11d98b)

## v1.2.0

* Able to require normally via `require('lokka-transport-http').Transport`;

## v1.1.0

* Handle 200 and 400 status code separately. Throw errors for others.
  * 200 is for success
  * 400 is for when there is a GraphQL Error

## v1.0.0

* Initial release


lib/__tests__/index.js
/* eslint max-len:0 */

import {describe, it} from 'mocha';
import {expect} from 'chai';
import LokkaHTTPTransport from '../';
import Schema from './sample_schema';
import express from 'express';
import expressGraphql from 'express-graphql';
import {promisify} from 'bluebird';
import {findAPortNotInUse} from 'portscanner';
const pickPort = () => promisify(findAPortNotInUse)(10000, 50000, '127.0.0.1');

describe('LokkaHTTPTransport', () => {
  it('should be able to require normally', () => {
    expect(require('../').Transport).to.be.equal(LokkaHTTPTransport);
  });

  describe('constructor()', () => {
    describe('without an endpoint', () => {
      it('should throw an error', () => {
        expect(() => new LokkaHTTPTransport()).to.throw(/endpoint is required!/);
      });
    });

    describe('with an endpoint', () => {
      it('should create an instance', () => {
        const transport = new LokkaHTTPTransport('/graphql');
        expect(transport.endpoint).to.be.equal('/graphql');
      });
    });
  });

  describe('_buildOptions', () => {
    it('should get the default options', () => {
      const transport = new LokkaHTTPTransport('/');
      const options = transport._buildOptions({aa: 10});

      expect(options).to.deep.equal({
        method: 'post',
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/json'
        },
        body: JSON.stringify({aa: 10}),
        credentials: 'include',
        mode: 'cors',
      });
    });

    it('should add custom headers', () => {
      const customOptions = {
        headers: {
          bb: 'hello'
        },
      };
      const transport = new LokkaHTTPTransport('/', customOptions);
      const options = transport._buildOptions({aa: 10});

      expect(options).to.deep.equal({
        method: 'post',
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/json',
          bb: 'hello'
        },
        body: JSON.stringify({aa: 10}),
        credentials: 'include',
        mode: 'cors',
      });
    });

    it('should add custom mode', () => {
      const customOptions = {
        mode: 'no-cors',
      };
      const transport = new LokkaHTTPTransport('/', customOptions);
      const options = transport._buildOptions({mode: 'same-origin'});

      expect(options).to.deep.equal({
        method: 'post',
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({mode: 'same-origin'}),
        credentials: 'include',
        mode: 'no-cors',
      });
    });

    describe('with unknown mode option', () => {
      it('should throw error', () => {
        expect(() => new LokkaHTTPTransport('/graphql', {mode: 'my-mode'})).to.throw('unknown mode: my-mode');
      });
    });
  });

  describe('send()', () => {
    describe('with a correct graphql query', () => {
      it('should return the result', async () => {
        const app = express();
        app.use('/graphql', expressGraphql({schema: Schema}));
        const port = await pickPort();
        await promisify(app.listen.bind(app))(port);

        const transport = new LokkaHTTPTransport(`http://localhost:${port}/graphql`);
        const result = await transport.send(`
          {
            echo(message: "Hello")
          }
        `);

        expect(result).to.be.deep.equal({
          echo: 'Echo: Hello'
        });
      });
    });

    describe('with an incorrect graphql query', () => {
      it('should return the error', async () => {
        const app = express();
        app.use('/graphql', expressGraphql({schema: Schema}));
        const port = await pickPort();
        await promisify(app.listen.bind(app))(port);

        const transport = new LokkaHTTPTransport(`http://localhost:${port}/graphql`);
        try {
          await transport.send(`
            {
              echo(messa: "Hello")
            }
          `);
          throw new Error('Some Other Error');
        } catch (err) {
          expect(err.message).to.match(/GraphQL Error:/);
          expect(err.rawError).to.be.instanceOf(Array);
        }
      });
    });

    describe('with an invalid status code', () => {
      it('should return the error', async () => {
        const app = express();
        const port = await pickPort();
        await promisify(app.listen.bind(app))(port);

        const transport = new LokkaHTTPTransport(`http://localhost:${port}/graphql`);
        try {
          await transport.send(`
            {
              echo(messa: "Hello")
            }
          `);
          throw new Error('Some Other Error');
        } catch (err) {
          expect(err.message).to.match(/Invalid status code: 404/);
        }
      });
    });
  });
});


lib/__tests__/sample_schema.js
import {
  GraphQLObjectType,
  GraphQLString,
  GraphQLSchema
} from 'graphql';

const schema = new GraphQLSchema({
  query: new GraphQLObjectType({
    name: 'MyQuery',
    fields: () => ({
      echo: {
        type: GraphQLString,
        args: {
          message: {type: GraphQLString}
        },
        resolve: (root, {message}) => {
          return `Echo: ${message}`;
        }
      }
    })
  })
});

export default schema;


lib/index.js
/* global fetch */
import LokkaTransport from 'lokka/transport';
// In some envionment like in ReactNative, we don't need fetch at all.
// Technically, this should be handle by 'isomorphic-fetch'.
// But it's not happening. So this is the fix

const isNode = new Function('try {return this===global;}catch(e){return false;}');

let fetchUrl;
if (typeof fetch === 'function') {
  // has a native fetch implementation
  fetchUrl = fetch;
} else if (isNode()) {
  // for Node.js
  fetchUrl = require('node-fetch');
  fetchUrl.Promise = Promise;
} else {
  // for the browser
  require('whatwg-fetch');
  fetchUrl = fetch;
}

// the default error handler
function handleErrors(errors, data) {
  const message = errors[0].message;
  const error = new Error(`GraphQL Error: ${message}`);
  error.rawError = errors;
  error.rawData = data;
  throw error;
}

export class Transport extends LokkaTransport {
  constructor(endpoint, options = {}) {
    if (!endpoint) {
      throw new Error('endpoint is required!');
    }

    super();
    this._httpOptions = {
      auth: options.auth,
      headers: options.headers || {},
      credentials: options.credentials,
    };
    this.endpoint = endpoint;
    this.handleErrors = options.handleErrors || handleErrors;
  }

  _buildOptions(payload) {
    const options = {
      method: 'post',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(payload),
      // To pass cookies to the server. (supports CORS as well)
      credentials: 'include',
    };

    // use delete property for backward compatibility
    if (this._httpOptions.credentials === false) {
      delete options.credentials;
    }

    Object.assign(options.headers, this._httpOptions.headers);
    return options;
  }

  send(query, variables, operationName) {
    const payload = {query, variables, operationName};
    const options = this._buildOptions(payload);

    return fetchUrl(this.endpoint, options).then(response => {
      // 200 is for success
      // 400 is for bad request
      if (response.status !== 200 && response.status !== 400) {
        throw new Error(`Invalid status code: ${response.status}`);
      }

      return response.json();
    }).then(({data, errors}) => {
      if (errors) {
        this.handleErrors(errors, data);
        return null;
      }

      return data;
    });
  }
}

export default Transport;
